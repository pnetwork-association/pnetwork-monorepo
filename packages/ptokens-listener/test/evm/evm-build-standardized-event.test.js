const { logs } = require('../mock/evm-logs')
const { getInterfaceFromEvent } = require('../../lib/evm/listener-evm')
const schemas = require('ptokens-schemas')
const { validation } = require('ptokens-utils')

describe('Event building for EVM', () => {
  describe('buildStandardizedEventFromEvmEvent', () => {
    beforeAll(() => {
      jest
        .useFakeTimers({ legacyFakeTimers: false })
        .setSystemTime(new Date('2023-03-14T16:00:00Z'))
    })

    it('Should build a standardized event object for a UserOperation event', async () => {
      const {
        buildStandardizedEvmEventObjectFromLog,
      } = require('../../lib/evm/evm-build-standardized-event')
      const chainId = '0xe15503e4'
      const eventName =
        'UserOperation(uint256 nonce,string destinationAccount,bytes4 destinationNetworkId,string underlyingAssetName,string underlyingAssetSymbol,uint256 underlyingAssetDecimals,address underlyingAssetTokenAddress,bytes4 underlyingAssetNetworkId,address assetTokenAddress,uint256 assetAmount,bytes userData,bytes32 optionsMask)'
      const eventLog = logs[1]
      const methodInterface = await getInterfaceFromEvent(eventName)
      const result = await buildStandardizedEvmEventObjectFromLog(
        chainId,
        methodInterface,
        eventLog
      )

      const expected = {
        [schemas.constants.SCHEMA_ID_KEY]:
          '0xf59387e05541d3d7690e8d54444179492f4b48a18106d94b1260e3560cf14acf',
        [schemas.constants.SCHEMA_STATUS_KEY]:
          schemas.db.enums.txStatus.DETECTED,
        [schemas.constants.SCHEMA_EVENT_NAME_KEY]:
          schemas.db.enums.eventNames.USER_OPERATION,

        [schemas.constants.SCHEMA_NONCE_KEY]: '103383',
        [schemas.constants.SCHEMA_ASSET_AMOUNT_KEY]: '100000000000000000000',
        [schemas.constants.SCHEMA_DESTINATION_ACCOUNT_KEY]:
          '0xa41657bf225F8Ec7E2010C89c3F084172948264D',
        [schemas.constants.SCHEMA_DESTINATION_NETWORK_ID_KEY]: '0xe15503e4',

        [schemas.constants.SCHEMA_FINAL_TX_HASH_KEY]: null,
        [schemas.constants.SCHEMA_FINAL_TX_TS_KEY]: null,
        [schemas.constants.SCHEMA_UNDERLYING_ASSET_NAME_KEY]: 'Token',
        [schemas.constants.SCHEMA_UNDERLYING_ASSET_SYMBOL_KEY]: 'TKN',
        [schemas.constants.SCHEMA_UNDERLYING_ASSET_DECIMALS_KEY]: 18,
        [schemas.constants.SCHEMA_UNDERLYING_ASSET_NETWORK_ID_KEY]:
          '0xe15503e4',
        [schemas.constants.SCHEMA_UNDERLYING_ASSET_TOKEN_ADDRESS_KEY]:
          '0x49a5D1CF92772328Ad70f51894FD632a14dF12C9',
        [schemas.constants.SCHEMA_OPTIONS_MASK]:
          '0x0000000000000000000000000000000000000000000000000000000000000000',
        [schemas.constants.SCHEMA_ORIGINATING_NETWORK_ID_KEY]: '0xe15503e4',
        [schemas.constants.SCHEMA_ORIGINATING_ADDRESS_KEY]: null,
        [schemas.constants.SCHEMA_ORIGINATING_BLOCK_HASH_KEY]:
          '0x509580316faa053e500a1cdba70359354d1f8d52ba7e8b7a90fc61e33a67f785',
        [schemas.constants.SCHEMA_ORIGINATING_TX_HASH_KEY]:
          '0x3f0a8aba3cc598f9e4fc851022044966238a9a9ab959c5130d9fa483204978f2',
        [schemas.constants.SCHEMA_PROPOSAL_TX_HASH_KEY]: null,
        [schemas.constants.SCHEMA_PROPOSAL_TS_KEY]: null,
        [schemas.constants.SCHEMA_ASSET_TOKEN_ADDRESS_KEY]:
          '0x49a5D1CF92772328Ad70f51894FD632a14dF12C9',
        [schemas.constants.SCHEMA_USER_DATA_KEY]: '0x',
        [schemas.constants.SCHEMA_WITNESSED_TS_KEY]: '2023-03-14T16:00:00.000Z',
      }
      await validation.validateJson(schemas.db.collections.events, expected)
      expect(result).toStrictEqual(expected)
    })
  })
})
