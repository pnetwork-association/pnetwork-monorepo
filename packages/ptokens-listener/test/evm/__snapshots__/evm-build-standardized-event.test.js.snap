// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event building for EVM buildStandardizedEventFromEvmEvent Should build a standardized event object for a OperationExecuted event 1 1`] = `
{
  "_id": "operationexecuted_0xe15503e4_0x79a1d514db0a65e136df72c0d40b553068ac455d0965a0486dcb6d1e71682435",
  "assetAmount": "99799999900",
  "assetTokenAddress": null,
  "blockHash": "0xfc7e05dc36287510a70fa64cd8c12ed2ae94aff985fd42bf164f19f97c28e38f",
  "destinationAccount": "0xa41657bf225F8Ec7E2010C89c3F084172948264D",
  "destinationNetworkId": "0xf9b459a1",
  "eventArgs": [
    [
      "0x6442fbb149dbc033aaac0b5b31322051d56c5de1db2c6d82b516eeb2929d2704",
      "0xce647bd765574c409f8115968644a3385411d82f4ca4ada1afe6cbc3c322b9e7",
      "0x0000000000000000000000000000000000000000000000000000000000000000",
      "24517",
      "18",
      "99799999900",
      "0",
      "100",
      "0",
      "0xdaacB0Ab6Fb34d24E8a67BfA14BF4D95D4C7aF92",
      "0xd41b1c5b",
      "0xf9b459a1",
      "0x00000000",
      "0x5aca268b",
      "0xa41657bf225f8ec7e2010c89c3f084172948264d",
      "0xa41657bf225F8Ec7E2010C89c3F084172948264D",
      "pNetwork Token",
      "PNT",
      "0x",
      false,
    ],
  ],
  "eventName": "OperationExecuted",
  "finalTransactionHash": null,
  "finalTransactionTimestamp": null,
  "forwardDestinationNetworkId": "0x00000000",
  "forwardNetworkFeeAssetAmount": "0",
  "isForProtocol": false,
  "networkFeeAssetAmount": "100",
  "networkId": "0xe15503e4",
  "nonce": "24517",
  "operationId": "0x73d96bb1831f73f4f47269f312b2131ca7a3becc4ccf8f53d5352575013c9378",
  "optionsMask": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "originatingAddress": "0xa41657bf225f8ec7e2010c89c3f084172948264d",
  "originatingBlockHash": "0x6442fbb149dbc033aaac0b5b31322051d56c5de1db2c6d82b516eeb2929d2704",
  "originatingNetworkId": "0xd41b1c5b",
  "originatingTransactionHash": "0xce647bd765574c409f8115968644a3385411d82f4ca4ada1afe6cbc3c322b9e7",
  "proposedTransactionHash": null,
  "proposedTransactionTimestamp": null,
  "status": "detected",
  "transactionHash": "0x79a1d514db0a65e136df72c0d40b553068ac455d0965a0486dcb6d1e71682435",
  "underlyingAssetDecimals": 18,
  "underlyingAssetName": "pNetwork Token",
  "underlyingAssetNetworkId": "0x5aca268b",
  "underlyingAssetSymbol": "PNT",
  "underlyingAssetTokenAddress": "0xdaacB0Ab6Fb34d24E8a67BfA14BF4D95D4C7aF92",
  "userData": "0x",
  "userDataProtocolFeeAssetAmount": "0",
  "witnessedTimestamp": "2023-03-14T16:00:00.000Z",
}
`;

exports[`Event building for EVM buildStandardizedEventFromEvmEvent Should build a standardized event object for a OperationQueued event 1 1`] = `
{
  "_id": "operationqueued_0xe15503e4_0x811297c76ff823b36c5e60952c68db3da8de1f57f1fc42dd4f0c9a323eb3ba8f",
  "assetAmount": "9979999900",
  "assetTokenAddress": null,
  "blockHash": "0x8726fbbfb4e1df49ea5fd83a552188233f7b90ea85b9d76ee5ddaf4ede5d16fb",
  "destinationAccount": "0xa41657bf225F8Ec7E2010C89c3F084172948264D",
  "destinationNetworkId": "0xf9b459a1",
  "eventArgs": [
    [
      "0xb9fe9822a3b786e95276cc41b2dab16c2d2e777b6193ed4d85382d39d392daba",
      "0x6024abc2fcf0a2c71ade50f5ce88a6799f6842fd449a4758984fe124b275b756",
      "0x0000000000000000000000000000000000000000000000000000000000000000",
      "24517",
      "18",
      "9979999900",
      "0",
      "100",
      "0",
      "0xdaacB0Ab6Fb34d24E8a67BfA14BF4D95D4C7aF92",
      "0xd41b1c5b",
      "0xf9b459a1",
      "0x00000000",
      "0x5aca268b",
      "0xa41657bf225f8ec7e2010c89c3f084172948264d",
      "0xa41657bf225F8Ec7E2010C89c3F084172948264D",
      "pNetwork Token",
      "PNT",
      "0x",
      false,
    ],
  ],
  "eventName": "OperationQueued",
  "finalTransactionHash": null,
  "finalTransactionTimestamp": null,
  "forwardDestinationNetworkId": "0x00000000",
  "forwardNetworkFeeAssetAmount": "0",
  "isForProtocol": false,
  "networkFeeAssetAmount": "100",
  "networkId": "0xe15503e4",
  "nonce": "24517",
  "operationId": "0x63abee74a702faf67a0f60b1fb775714f7c2b19703bc5e39fac437023d9b56c4",
  "optionsMask": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "originatingAddress": "0xa41657bf225f8ec7e2010c89c3f084172948264d",
  "originatingBlockHash": "0xb9fe9822a3b786e95276cc41b2dab16c2d2e777b6193ed4d85382d39d392daba",
  "originatingNetworkId": "0xd41b1c5b",
  "originatingTransactionHash": "0x6024abc2fcf0a2c71ade50f5ce88a6799f6842fd449a4758984fe124b275b756",
  "proposedTransactionHash": null,
  "proposedTransactionTimestamp": null,
  "status": "proposed",
  "transactionHash": "0x811297c76ff823b36c5e60952c68db3da8de1f57f1fc42dd4f0c9a323eb3ba8f",
  "underlyingAssetDecimals": 18,
  "underlyingAssetName": "pNetwork Token",
  "underlyingAssetNetworkId": "0x5aca268b",
  "underlyingAssetSymbol": "PNT",
  "underlyingAssetTokenAddress": "0xdaacB0Ab6Fb34d24E8a67BfA14BF4D95D4C7aF92",
  "userData": "0x",
  "userDataProtocolFeeAssetAmount": "0",
  "witnessedTimestamp": "2023-03-14T16:00:00.000Z",
}
`;

exports[`Event building for EVM buildStandardizedEventFromEvmEvent Should build a standardized event object for a UserOperation event 1 1`] = `
{
  "_id": "useroperation_0x5aca268b_0xa79079c344c44721e5040643d6aaa375d2a97d5a28b0a57e4cb4dc0f84e7409d",
  "assetAmount": "1230000000000000",
  "assetTokenAddress": "0xdaacB0Ab6Fb34d24E8a67BfA14BF4D95D4C7aF92",
  "blockHash": "0x20df4aec83af7e41dd37b13effaeda642e180b02da38e81c662a0a2b362f467b",
  "destinationAccount": "0xADA2de876567a06eD79b0B29ae6aB2e142129E51",
  "destinationNetworkId": "0xd41b1c5b",
  "eventArgs": [
    "103237",
    "0xada2de876567a06ed79b0b29ae6ab2e142129e51",
    "0xADA2de876567a06eD79b0B29ae6aB2e142129E51",
    "0xd41b1c5b",
    "pNetwork Token",
    "PNT",
    "18",
    "0xdaacB0Ab6Fb34d24E8a67BfA14BF4D95D4C7aF92",
    "0x5aca268b",
    "0xdaacB0Ab6Fb34d24E8a67BfA14BF4D95D4C7aF92",
    "1230000000000000",
    "0",
    "100",
    "100",
    "0xd41b1c5b",
    "0x5aca268b",
    "0x",
    "0x0000000000000000000000000000000000000000000000000000000000000000",
    false,
  ],
  "eventName": "UserOperation",
  "finalTransactionHash": null,
  "finalTransactionTimestamp": null,
  "forwardDestinationNetworkId": "0xd41b1c5b",
  "forwardNetworkFeeAssetAmount": "100",
  "isForProtocol": false,
  "networkFeeAssetAmount": "100",
  "networkId": "0x5aca268b",
  "nonce": "103237",
  "operationId": "0xc63cad3a15cd83a264c227dabb25107343e0e82fa0b9083bbbbed93aa3ad94f1",
  "optionsMask": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "originatingAddress": "0xada2de876567a06ed79b0b29ae6ab2e142129e51",
  "originatingBlockHash": null,
  "originatingNetworkId": "0x5aca268b",
  "originatingTransactionHash": null,
  "proposedTransactionHash": null,
  "proposedTransactionTimestamp": null,
  "status": "detected",
  "transactionHash": "0xa79079c344c44721e5040643d6aaa375d2a97d5a28b0a57e4cb4dc0f84e7409d",
  "underlyingAssetDecimals": 18,
  "underlyingAssetName": "pNetwork Token",
  "underlyingAssetNetworkId": "0x5aca268b",
  "underlyingAssetSymbol": "PNT",
  "underlyingAssetTokenAddress": "0xdaacB0Ab6Fb34d24E8a67BfA14BF4D95D4C7aF92",
  "userData": "0x",
  "userDataProtocolFeeAssetAmount": "0",
  "witnessedTimestamp": "2023-03-14T16:00:00.000Z",
}
`;
